### DISCUSSION TITLE
Nested components JS creates multiple listeners

### DISCUSSION THREAD

**Discussion by polycode-nz** [livewire/livewire#8040] [1. Help]
Livewire V3.0

How are we meant to include a nested components JS?

Every way I do it, it seems to multiple the events/listeners  in the JS by the number of times the nested component is run. 

There's a few posts about this already:
https://github.com/livewire/livewire/discussions/7366
https://github.com/livewire/livewire/discussions/6662
https://github.com/livewire/livewire/discussions/6910
https://github.com/livewire/livewire/discussions/6927

@push doesn't work in the nested components blade.

I have tried these:

```
@once
     @script
        <script data-navigate-once>
```

With and without @once, @script and data-navigate-once.

I have tried it in an x-int on the nested components outer div.

All of it results in multiple event handlers/listeners being run.

Is there a definitive way to achieve modularised javascript (as in keeping a components JS to itself) and solving this problem?

**Comment by polycode-nz** [comment]
No clues?

Best I can do so far is set a var in the parent.

In the parents blade I have:

```
 <script>
    var hasRunBefore = false;
 </script>
```

In the child components blade I have:

```
@script
     <script>
        if ( ! hasRunBefore)
        {
            hasRunBefore = true;

            $body.on('click', '.js-read-btn', function(){
                // removed brevity
                Livewire.dispatch('readJob', {id: $this.data('id')});
            });

            Livewire.on('openReadJobDrawer', (event) => {
                // so stuff
            }
        }
    </script>
@endscript
```

That works and stops the multiplication of the two listeners.  Should that be necessary?




I see this example in the docs: https://livewire.laravel.com/docs/events#listening-for-livewire-events-from-global-javascript

```
<script>
    document.addEventListener('livewire:init', () => {
        let cleanup = Livewire.on('post-created', (event) => {
            //
        });
 
        // Calling "cleanup()" will un-register the above event listener...
        cleanup();
    });
</script>


```

Which didn't work at all. The `openReadJobDrawer` listener never runs.

**Comment by joshhanley** [comment]
@polycode-nz any chance you could submit a failing test PR that demonstrates the issue so we can investigate? Thanks!

**Reply by polycode-nz** [reply]
Jason Beggs has taken a look and it's just how it is.

This nested component is in a modal that has it's dom added and removed from the parent component dom. When it get's re-added on the second and subsequent opens of the parent modal the x-init of the nested component runs again and re-adds the click handlers and listeners.

It makes sense.. but I don't have a fix
